dashboard/utils_test.py

```diff
@@ -0,0 +1,395 @@
+import os
+from datetime import datetime, timedelta
+
+import pytz
+from django.core.exceptions import ImproperlyConfigured
+from django.test import TestCase
+from edx_api.enrollments.models import Enrollments
+from edx_api.certificates.models import Certificate, Certificates
+from edx_api.grades.models import CurrentGrade, CurrentGrades
+
+from courses.factories import ProgramFactory, CourseRunFactory
+from dashboard.utils import MMTrack
+from ecommerce.factories import CoursePriceFactory, LineFactory, OrderFactory
+from ecommerce.models import Order
+from financialaid.factories import TierProgramFactory, FinancialAidFactory
+from profiles.factories import UserFactory
+
+
+class MMTrackTest(TestCase):
+    """
+    Tests for the MMTrack class
+    """
+
+    @classmethod
+    def setUpTestData(cls):
+        super().setUpTestData()
+        # create an user
+        cls.user = UserFactory.create()
+
+        # create Enrollments, Certificates, CurrentGrades
+        with open(os.path.join(os.path.dirname(__file__),
+                               'fixtures/user_enrollments.json')) as file_obj:
+            cls.enrollments_json = json.loads(file_obj.read())
+        cls.enrollments = Enrollments(cls.enrollments_json)
+
+        with open(os.path.join(os.path.dirname(__file__),
+                               'fixtures/certificates.json')) as file_obj:
+            certificates_json = json.loads(file_obj.read())
+        cls.certificates = Certificates([Certificate(cert_json) for cert_json in certificates_json])
+
+        with open(os.path.join(os.path.dirname(__file__),
+                               'fixtures/current_grades.json')) as file_obj:
+            current_grades_json = json.loads(file_obj.read())
+        cls.current_grades = CurrentGrades([CurrentGrade(grade_json) for grade_json in current_grades_json])
+
+        # create the programs
+        cls.program = ProgramFactory.create(live=True, financial_aid_availability=False)
+        cls.program_financial_aid = ProgramFactory.create(live=True, financial_aid_availability=True)
+
+        # create course runs for the normal program
+        for course_key in ["course-v1:edX+DemoX+Demo_Course", "course-v1:MITx+8.MechCX+2014_T1", '']:
+            CourseRunFactory.create(
+                program=cls.program,
+                edx_course_key=course_key
+            )
+        # and the program with financial aid
+        cls.now = datetime.now(pytz.utc)
+        cls.end_date = cls.now - timedelta(weeks=45)
+        cls.crun_fa = CourseRunFactory.create(
+            program=cls.program_financial_aid,
+            start_date=cls.now-timedelta(weeks=52),
+            end_date=cls.end_date,
+            enrollment_start=cls.now-timedelta(weeks=62),
+            enrollment_end=cls.now-timedelta(weeks=53),
+            edx_course_key="course-v1:odl+FOO101+CR-FALL15"
+        )
+        CourseRunFactory.create(
+            program=cls.program_financial_aid,
+            edx_course_key=""
```

Comment:
We do not want to handle course run with empty `edx_course_key`, meaning runs that have not a correspondent run on edX